lib/broadway_dashboard/metrics.ex

```diff
@@ -85,31 +119,24 @@ defmodule BroadwayDashboard.Metrics do
   end
 
   @impl true
-  def handle_call({:ensure_counters_restarted, pipeline}, _, state) do
-    if Map.get(state.listeners, pipeline) do
-      Counters.start!(pipeline)
-    end
+  def handle_call(:ensure_counters_restarted, _, state) do
+    Counters.start!(state.pipeline)
 
     {:reply, :ok, state}
   end
 
   @impl true
-  def handle_call({:listen, parent, pipeline}, _, state) do
-    case Process.whereis(pipeline) do
+  def handle_call({:listen, parent}, _, state) do
+    case Process.whereis(state.pipeline) do
       pid when is_pid(pid) ->
         ref = Process.monitor(parent)
+        refs = MapSet.put(state.refs, ref)
 
-        listeners = Listeners.add(state.listeners, pipeline, parent)
+        listeners = MapSet.put(state.listeners, parent)
```

Comment:
Instead of two mapsets, I think you can have a single map where refs are keys and listeners are values. Then to send a message, you will message all listeners (values).